disperser/controller/dispatcher.go

```diff
@@ -152,15 +229,22 @@ func (d *Dispatcher) HandleBatch(ctx context.Context) (chan core.SigningMessage,
 		d.metrics.reportHandleBatchLatency(time.Since(start))
 	}()
 
-	currentBlockNumber, err := d.chainState.GetCurrentBlockNumber(ctx)
-	if err != nil {
-		return nil, nil, fmt.Errorf("failed to get current block number: %w", err)
+	// Get the cached onchain state. If it is outdated, we need to manually fetch it. If we fail to fetch it
+	// we will not be able to make a new batch.
+	cachedOnChainState := d.cachedOnchainState.Load().(*CachedOnchainState)
+	if time.Since(cachedOnChainState.LastRefreshed) > d.OnchainStateRefreshInterval {
+		d.logger.Warn("cached onchain state is outdated, manually fetching it")
+		err := d.refreshOnchainState(ctx)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to refresh onchain state: %w", err)
+		}
+		// reload the fresh snapshot
+		cachedOnChainState = d.cachedOnchainState.Load().(*CachedOnchainState)
 	}
-	referenceBlockNumber := uint64(currentBlockNumber) - d.FinalizationBlockDelay
 
 	// Get a batch of blobs to dispatch
 	// This also writes a batch header and blob inclusion info for each blob in metadata store
-	batchData, err := d.NewBatch(ctx, referenceBlockNumber)
+	batchData, err := d.NewBatch(ctx, cachedOnChainState)
```

Comment:
`cachedOnChainState` is only used by the `NewBatch()` method. Would it be possible to fetch the chain state inside `NewBatch()` instead of passing it in? Perhaps you could even encapsulate the block of code that fetches `cachedOnChainState` into a helper method.

`HandleBatch()` is already a very long and complicated method, and so extracting code helps to reduce its complexity.